### ISSUE TITLE
When running a workload with a single control plane node the load balancers take 15 mins to provision

### ISSUE THREAD

**Issue by jsturtevant** [kubernetes-sigs/cluster-api-provider-azure#857]
/kind bug

**status** (as of 6/10/21):

- fixed in out of tree cloud provider in https://github.com/kubernetes-sigs/cloud-provider-azure/pull/537 (included in v0.7.3+)
- cherry picked into in tree cloud provider in https://github.com/kubernetes/kubernetes/pull/100110 (included in k8s 1.22+)

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
When running a workload with a single control plane node the load balancers take 15 mins to provision.  

Add the following to the `Creating a single control-plane cluster with 1 worker node` e2e test after cluster creation:

```
AzureLBSpec(ctx, func() AzureLBSpecInput {
					return AzureLBSpecInput{
						BootstrapClusterProxy: bootstrapClusterProxy,
						Namespace:             namespace,
						ClusterName:           clusterName,
						SkipCleanup:           skipCleanup,
					}
				})
```

Run the e2e test and the test will fail:

```
./scripts/ci-e2e.sh

## commented out

Workload cluster creation                                                                                                                                                                           
[1] /home/jstur/projects/cluster-api-provider-azure/test/e2e/azure_test.go:36                                                                                                                           
[1]   Creating a single control-plane cluster                                                                                                                                                           
[1]   /home/jstur/projects/cluster-api-provider-azure/test/e2e/azure_test.go:71                                                                                                                         
[1]     With 1 worker node [It]                                                                                                                                                                         
[1]     /home/jstur/projects/cluster-api-provider-azure/test/e2e/azure_test.go:72                                                                                                                       
[1]                                                                                                                                                                                                     
[1]     Timed out after 180.000s.                                                                                                                                                                       
[1]     Service default/ingress-nginx-ilb failed to get an IP for LoadBalancer.Ingress                                                                                                                  
[1]     Expected                                                                                                                                                                                        
[1]         <bool>: false                                                                                                                                                                               
[1]     to be true                                                                                                                                                                                      
[1]                                                                                                                                                                                                     
[1]     /home/jstur/projects/cluster-api-provider-azure/test/e2e/helpers.go:97  
```

If you connect to the workload cluster you will see the service with the load balancer is there and after 15 mins will provision. Subsequent services with load balancers will provision quickly.  The logs of the controller manager will contain:

```
E0802 23:30:42.090814       1 azure_vmss.go:1116] EnsureHostInPool(default/ingress-nginx-ilb): backendPoolID(/subscriptions/b9d9436a-0c07-4fe8-b779-2c1030bd7997/resourceGroups/capz-e2e-72fll1/providers/Microsoft.Network/loadBalancers/capz-e2e-72fll1-internal/backendAddressPools/capz-e2e-72fll1) - failed to ensure host in pool: "not a vmss instance"
```

**What did you expect to happen:**
That the tests should be able to provision a workload cluster and pass a e2e test that creates a load balancer.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

This is related to https://github.com/kubernetes-sigs/cloud-provider-azure/issues/338

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by alexeldeib** [comment]
Guess this is another code path related to not using Availability Sets? We should probably consider that as mitigation. Anything that tries to look up IDs will fail with our current setup. It's hard to track down all the places individually.

**Comment by jsturtevant** [comment]
It is related but not the root cause, The root cause is the cache used in the controller-manager.  I provided more details in  kubernetes-sigs/cloud-provider-azure#363.  

I believe this could cause delays in a customer scenario were a node is added after the fact causing a delay of the LB provisioning due the fact that the cache doesn't know about the node.

**Comment by jsturtevant** [comment]
I ran into this again trying to set up a single control plan test for windows.  This appears to be an issue only in the VMAS sceario.  There is VMSS tests that is a single node: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/cb486c3d7b26dc84ad5156fa97773ccb97578ebe/test/e2e/azure_test.go#L260-L261

**Comment by CecileRobertMichon** [comment]
@devigned @lastcoolnameleft the current version of external-cloud-provider we're using in the example template in CAPZ v0.4 doesn't have the fix yet unfortunately, the PR to bump the version (#1323) and enable the test that validates this behavior was blocked by another regression in cloud-provider, which is now released. You can work around it for now by editing your template to use version v0.7.4+ of cloud-provider, until we update the reference template.

The in-tree fix will be in k8s 1.22+.

Regarding using out of tree by default, v1.0.0 of out-of-tree provider just got released so it might be a good time that, tracking in #715

**Comment by CecileRobertMichon** [comment]
I think we can close this now. This was fixed in the external cloud provider v0.7.4+ and k8s 1.22+.

/close
